### ISSUE TITLE
Ivy runscript changes

### ISSUE THREAD

**Issue by Zidane** [SpongePowered/Sponge#326]
This is meant for @Mumfrey 

After reviewing how people generally use the Ivy runscript (which is still running great), I'm not seeing a real value in keeping the run command or install.

Reasoning is dependency upgrades (including Forge). People are getting in the habit of using "run" and wondering why Sponge breaks. While the log does make this PAINFULLY clear that yes, your Forge version is out of date, there is some misconception that run would handle this for you. In reality, run takes the caches and "runs" with it.

I don't see a case in-which a server admin would want to run with the current caches on a new Sponge build. 9/10 times when this happens its a mistake and they meant to run "go" or "install + run". To me it seems like we should cut out the possibility of failure and simply have it check for dependencies each launch.

Thoughts?

**Comment by Mumfrey** [comment]
There was some thought put into the whole `install`, `run`, `go` thing, and perhaps - based on your observations - some change is required. However let me explain the original thinking first so that it's clear why those options exist in the first place.

There is certainly some stigma attached to "dep-loading" with no user intervention, that is to say a supposedly stand-alone jar simply fetching its dependencies with no user intervention whatsoever, and effectively adding binaries to your computer which you did *not* implicitly install. This is especially true on the client side, maybe slightly less so on a server but still a valid concern.

`install` makes it patently clear that fetching of dependencies is what it's about: it will get the things which make the other things do the thing. `run` on the other hand carries with it an explicit contract of **not** fetching the things to do the thing, with the implication that `install` is required to do the thing, putting fetching of things in the hands of the end user. This is designed to make it clear - at the front end - that `install` requires a net connection and **will download things** and `run` does not and **will not download things**.

This separation is partly for reasons of transparency and partly for reasons of just pure speed - the depload check is just patently *not* required on every server start and someone who wants the option to skip it should have that option available to them.

I think, pretty much, that the existence of the three behaviours is not really in question here, because they all make sense. What **is** the live issue is basically what users understand those behaviours to mean, and potentially putting some mechanisms in place to inhibit, or at least resolve any misunderstandings.

The focus of this - to my mind - is the `run` behaviour. If what you're describing represents the majority of the problems people are encountering, then it's entirely feasible to write some logic which effectively adds some cushioning around the functionality of `run`, and prompts users to re-run `install` when necessary.

Other than that, I think we'd be trying to solve a problem which is going to exist until the end of time, which is - plainly - that **users who don't read the instructions are basically always going to have a bad time**. This is what it boils down to, and there is pretty much nothing we can ever do to fix that completely, all we can do is sprinkle airbags and soft cushions around the areas they're most likely to land. In this case, adding some additional logic to `run` to handle the scenarios you describe would seem to be the path of least resistance, because if we make it our mission to protect people from themselves we will always fail, and if we take away features which have an explicit (and perfectly valid) use-case then we just fail as well.

tl;dr - *let's fix `run` so that users confused by it aren't hurt by it*

**Comment by Zidane** [comment]
Then we need to cushion it cause any time that users run the ivy script and it fails due to a misunderstanding that translates into a misconception that Sponge isn't stable and is "dangerous to use". I'm not saying we can make this fool proof (and we certainly agree on that front).

Last thing we need is an undeserved reputation.

**Comment by Mumfrey** [comment]
Precisely.

**Comment by Zidane** [comment]
Looks like this has worked itself out over time.
